src/net/download.rs

```diff
@@ -0,0 +1,1859 @@
+}
+
+impl BlockDownloader {
+    pub fn new(dns_timeout: u128, max_inflight_requests: u64) -> BlockDownloader {
+        BlockDownloader {
+            state: BlockDownloaderState::GetBlocksBegin,
+
+            block_sortition_height: 0,
+            microblock_sortition_height: 0,
+            next_block_sortition_height: 0,
+            next_microblock_sortition_height: 0,
+
+            num_blocks_downloaded: 0,
+            num_microblocks_downloaded: 0,
+            empty_block_download_passes: 0,
+            empty_microblock_download_passes: 0,
+            finished_scan_at: 0,
+            last_inv_update_at: 0,
+
+            max_inflight_requests: max_inflight_requests,
+            blocks_to_try: HashMap::new(),
+            microblocks_to_try: HashMap::new(),
+
+            parsed_urls: HashMap::new(),
+            dns_lookups: HashMap::new(),
+            dns_timeout: dns_timeout,
+
+            getblock_requests: HashMap::new(),
+            getmicroblocks_requests: HashMap::new(),
+            blocks: HashMap::new(),
+            microblocks: HashMap::new(),
+
+            dead_peers: vec![],
+            broken_peers: vec![],
+            broken_neighbors: vec![],
+        }
+    }
+
+    pub fn reset(&mut self) -> () {
+        self.state = BlockDownloaderState::DNSLookupBegin;
+
+        self.dns_lookups.clear();
+        self.parsed_urls.clear();
+
+        self.getblock_requests.clear();
+        self.getmicroblocks_requests.clear();
+        self.blocks_to_try.clear();
+        self.microblocks_to_try.clear();
+        self.blocks.clear();
+        self.microblocks.clear();
+
+        self.dead_peers.clear();
+        self.broken_peers.clear();
+        self.broken_neighbors.clear();
+
+        // perserve sortition height
+        // preserve download accounting
+    }
+
+    pub fn restart_scan(&mut self) -> () {
+        // prepare to restart a full-chain scan for block downloads
+        self.block_sortition_height = 0;
+        self.microblock_sortition_height = 0;
+        self.next_block_sortition_height = 0;
+        self.next_microblock_sortition_height = 0;
+        self.empty_block_download_passes = 0;
+        self.empty_microblock_download_passes = 0;
+    }
+
+    pub fn dns_lookups_begin(&mut self, dns_client: &mut DNSClient, mut urls: Vec<UrlString>) -> Result<(), net_error> {
+        assert_eq!(self.state, BlockDownloaderState::DNSLookupBegin);
+
+        self.dns_lookups.clear();
+        for url_str in urls.drain(..) {
+            let url = url_str.parse_to_block_url()?;        // NOTE: should always succeed, since a UrlString shouldn't decode unless it's a valid URL
+            let port = match url.port_or_known_default() {
+                Some(p) => p,
+                None => {
+                    warn!("Unsupported URL {:?}: unknown port", &url);
+                    continue;
+                }
+            };
+            match url.host() {
```

Comment:
Should we also abort on non `http/https` schemes?

Reply:
We shouldn't abort on any network code path, since bad peers can submit malformed URLs.  If the peer submits a non-http URL, then it won't be used (and in my relay PR, the peer will get disconnected and banned for submitting a bad URL).